### ISSUE TITLE
In C#, add an additional parameter to `[Export]` to customize the exported property's name on the property list.

### ISSUE THREAD

**Issue by MagdielM** [godotengine/godot-proposals#3510]
### Describe the project you are working on

A 3D prototype with classes that export a considerable amount of properties.

### Describe the problem or limitation you are having in your project

Currently, the process to categorize exported script properties in the inspector involves overriding `_get()`, `_set()`, and `_get_property_list()`, ensuring all variables that should be exported are accounted for in each override, as well as making the script in question a tool script, which adds another consideration to keep track of when writing scripts (`!Engine.EditorHint`). This process can be quite cumbersome, as shown by #3451 and #1255. However, `export` already exposes much of the functionality you get from adding property entries manually via `_get_property_list()`, with the only prominent exception being the property's name, since it takes the name in the property declaration.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Godot already has magic separators (`/`) to use in place of exporting properties with `PROPERTY_USAGE_CATEGORY`, but these can't be used in the variable name in either GDScript or C#. I'm hesitant to suggest this for GDScript given all the 2.0 changes, but for C#, the `Export` attribute could take on an additional optional parameter to customize the property's display name.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

An additional overload for the `[Export]` attribute:
```cs
public ExportAttribute(PropertyHint hint = PropertyHint.None, string hintString = "", string displayName = null);
```
where this:
```cs
[Export(PropertyHint.Range, "0, 20", "SomeCategory/SomeProperty")]
public int Property { get; set; }
```
would serve as syntax sugar for this:
```cs
public int Property { get; set; }

public override object _Get(string property)
{
   if (property == "SomeCategory/SomeProperty")
   {
      return Property;
   }
   else
   {
      return base._Get(property);
   }
}

public override bool _Set(string property, object value)
{
   if (property == "SomeCategory/SomeProperty")
   {
      Property = (int)value;
      return true;
   }
   else
   {
      return base._Set(property, value);
   }
}

public override void _GetPropertyList()
{
   return new Godot.Collections.Array
   {
      new Godot.Collections.Dictionary
      {
         { "name", "SomeCategory/SomeProperty" },
         { "type", Variant.Type.Int },
         { "hint", PropertyHint.Range },
         { "hint_string", "0, 20" },
         { "usage", PropertyUsageFlags.Default }
      }
   }
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can be worked around with very many lines of script and the tedium of making sure to wrap any code you don't want to run in the editor around `if (!Engine.EditorHint)`.

### Is there a reason why this should be core and not an add-on in the asset library?

Addons can't modify the way Godot's C# attributes work.

**Comment by andy-noisyduck** [comment]
A similar alternative would be to keep the handling of the property name as it is now, but allow a category string to be passed via the attribute.

**Comment by MagdielM** [comment]
> A similar alternative would be to keep the handling of the property name as it is now, but allow a category string to be passed via the attribute.

True, but letting you customize the name also opens the door for displaying things with names that variables can't have, like names starting with numbers ("2DSomethingOrOther").

**Comment by pycbouh** [comment]
Just a clarification, because you called it `displayName` in the example:

Properties can't have a display name. You can change the name of the property itself, but its display name is exactly what the property name is with additional formatting optionally applied. What you suggest here is a way to define the exposed property under a different name than the backing internal field.

PS. It would also make sense to not limit this to just Mono.

**Comment by MagdielM** [comment]
> Properties can't have a display name. You can change the name of the property itself, but its display name is exactly what the property name is with additional formatting optionally applied. What you suggest here is a way to define the exposed property under a different name than the backing internal field.

Oh, so this is why we override `_get()` and `_set()`. Interesting.

Also yeah I thought about making it a general proposal, but I haven't looked into how the new `@export` annotation is different from the keyword, so I'm not in a position to suggest changes to how it's handled in GDScript.

**Comment by GeorgeS2019** [comment]
This PR: [**Add C# resource export**](https://github.com/godotengine/godot/pull/63126) could be relevant

**Comment by willnationsdev** [comment]
My C# resource export PR currently does not account for all of the different arguments, and if you use `[Export("SomeName")`, then that becomes short for `[Export(hint=PROPERTY_HINT_RESOURCE_TYPE,hintString="SomeName")]`. It wouldn't be hard to add the extra simplified constructor as an option though.

**Comment by fhd** [comment]
I'd love to have this mainly to stay consistent between properties exported from GDScript and C#. Right now my options are:

1. Use lower_case_with_underscores for exported properties in C#.
2. Use PascalCase for exported properties in GDScript.
3. Have PascalCase and lower_case_with_underscores depending on whether a script was written in C# or GDScript (i.e. give up on consistent naming).

I'm currently going with option 1, but it's a bit irritating.

**Comment by natrist** [comment]
Is there any progress on this issue?

**Comment by a0405u** [comment]
One huge issue that this proposal may solve is organizing things in huge classes, like complex character controllers, that are usually have pretty verbose names that are unable to be displayed fully in the editor.

Godot does not support any nested global resource classes (I hope it will someday) so there is no way to organize properties for huge nodes without having full long names using multiple classes and their instances. The solution for this is splitting the implementation to multiple scripts but it is not a good way to organize things in that case since it separates things that should not be separated but only organized. It also introduces additional implementation difficulties, such as the need to link classes with a separate implementation. Using nodes instead of inner resource classes are not the solution for all scenarios, since it still separates things that are only needed to be organized.

Dictionaries may work to wrap the parameters but their exported UI is not very convenient, they also have some influence on performance and can't be indexed through dots on C#, so it creates additional issues on the programming side. It is better to have a solution that does not affect how the game was programmed.

The only working way to organize the editor is to make groups, but in that case the names of the variables are still verbose, since there is no organization in the actual code. So property names are still displayed badly in the inspector. Having the ability to customize names for the properties that are grouped in the editor is necessary to get rid of unnecessary name prefixes that are used to differentiate properties in code.

**Comment by tp1415926535** [comment]
Opened an engine PR for this here: godotengine/godot#119326

This implementation only customizes the Inspector display label for C# exports. It does not change the underlying property name, serialization key, or get/set behavior.

**Example usage:**

```csharp
[Export]
[ExportName("Short Name")]
public string LongPropertyNameInTheCsharpScript { get; set; }
```
This shows `Short Name` in the Inspector while keeping the actual exported property name as `LongPropertyNameInTheCSharpScript`.


**Rationale:**
This is introduced as a separate attribute instead of adding more arguments to `[Export]`.
`[Export]` already carries export and Inspector editing metadata such as PropertyHint and HintString. `ExportName` addresses a different concern: overriding the Inspector label for an exported member without changing the underlying property name used by the engine.
Keeping it separate avoids overloading `[Export]` further and keeps the API clearer about what affects editing behavior versus what only affects presentation.
